packages/vite/src/node/ssr/runtime/runtime.ts

```diff
@@ -0,0 +1,327 @@
+import type { ViteHotContext } from 'types/hot'
+import { HMRClient, HMRContext } from '../../../shared/hmr'
+import { ModuleCacheMap } from './moduleCache'
+import type {
+  FetchResult,
+  ImportMetaEnv,
+  ModuleCache,
+  ResolvedResult,
+  SSRImportMetadata,
+  ViteModuleRunner,
+  ViteRuntimeImportMeta,
+  ViteRuntimeModuleContext,
+  ViteServerClientOptions,
+} from './types'
+import {
+  cleanUrl,
+  createImportMetaEnvProxy,
+  isPrimitive,
+  posixDirname,
+  posixPathToFileHref,
+  posixResolve,
+  unwrapId,
+} from './utils'
+import {
+  ssrDynamicImportKey,
+  ssrExportAllKey,
+  ssrImportKey,
+  ssrImportMetaKey,
+  ssrModuleExportsKey,
+} from './constants'
+import { ViteRuntimeHMRLogger, ViteRuntimeSilentHMRLogger } from './hmrLogger'
+
+interface ViteRuntimeDebugger {
+  (formatter: unknown, ...args: unknown[]): void
+}
+
+export class ViteRuntime {
+  /**
+   * Holds the cache of modules
+   * Keys of the map are ids
+   */
+  public moduleCache: ModuleCacheMap
+  public hmrClient?: HMRClient
+  public entrypoints = new Set<string>()
+
+  private idToFileMap = new Map<string, string>()
+  private envProxy: ImportMetaEnv
+
+  constructor(
+    public options: ViteServerClientOptions,
+    public runner: ViteModuleRunner,
+    private debug?: ViteRuntimeDebugger,
+  ) {
+    this.moduleCache = options.moduleCache ?? new ModuleCacheMap(options.root)
+    this.envProxy = createImportMetaEnvProxy(options.environmentVariables)
+    if (typeof options.hmr === 'object') {
+      this.hmrClient = new HMRClient(
+        options.hmr.logger === false
+          ? new ViteRuntimeSilentHMRLogger()
+          : options.hmr.logger || new ViteRuntimeHMRLogger(),
+        options.hmr.connection,
+        ({ acceptedPath, ssrInvalidates }) => {
+          this.moduleCache.delete(acceptedPath)
+          ssrInvalidates.forEach((id) => this.moduleCache.delete(id))
+          return this.executeUrl(acceptedPath, false)
+        },
+      )
+    }
+  }
+
+  /**
+   * URL to execute. Should be in the format of "/file.ts" like in the browser
+   */
+  public async executeUrl<T = any>(
+    url: string,
+    entrypoint = false,
+  ): Promise<T> {
+    const fetchedModule = await this.cachedModule(url)
+    return await this.cachedRequest(url, fetchedModule, [], { entrypoint })
+  }
+
+  /**
+   * Entrypoiont URLs to execute. Expects server URLs in the format of "/file.ts" like in the browser
+   * In the case of a full reload triggered by HMR, these are the modules that will be reloaded
+   */
+  public async executeEntrypoints<T extends any[]>(urls: string[]): Promise<T> {
+    const results: any[] = []
+    for (const entrypoint of urls) {
+      results.push(await this.executeUrl(entrypoint, true))
+    }
+    return results as T
+  }
```

Comment:
Is there a solid reason for this function to be part of the API? If we think it is useful to later on speed up things somehow, then I think it is good to have it, but if not, we could document this as a code snippet and let people use `executeUrl` directly.
About the name, should `executeUrl` be called `executeRequest` to be aligned with `cacheRequest`, `transformRequest`, `warmupRequest`?

Reply:
`executeUrl` is named like that because it accepts a url (it's `executeId` in vite-node, but I changed the meaning of the first parameter to have it closer to how Vite SSR works). All other functions accept several internal parameters. I think calling it a request in a public API would be wrong.

The reason why `executeEntrypoints` exists is that it's a shorthand for `executeUrl(..., true)` - this is an analog for `ssrLoadModule` where you call it only once. I am not sure of what performance benefits you are talking about here, to be honest, so I cannot comment on that. The reason why we have an "entry point" entity is that we need to know a module that would be reloaded if HMR triggers `full-reload`.

Reply:
I am actually not even sure if we should expose "entrypoint", maybe we should just treat any `executeUrl` call as an entrypoint 🤷🏻  - it depends on the usage

Reply:
I think what makes noise to me is the plural. So maybe it could be `executeEntrypoint` if you want a explicit way to do `executeUrl(..., true)`, and let the user do the iteration. In some cases it could look simpler from the caller side when there is a single entry point for example.

Ok on the `executeUrl` naming, thanks for the explanation 👍🏼

Reply:
Updated entrypoint logic